use avx2 and fma in adagrad

jspark1105 opened the pull request:
Summary: Using FMA should be faster and provide (slight) accuracy benefits.

Test Plan:
buck test caffe2/caffe2/fb/dper/layer_models/tests/split_1:sparse_nn_test_4

gives the following errors and this may tell us that these tests are either numerically too sensitive or the tolerance range is too small. For test_beta_loss, I guess using fma actually improved the accuracy.

```
> ======================================================================
> FAIL: test_cat_cat_embedding_arch_with_transfter_on_last (caffe2.caffe2.fb.dper.layer_models.tests.split_1.sparse_nn_test_4.SparseNNTest4)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/data/users/jongsoo/fbsource/fbcode/buck-out/dev/gen/caffe2/caffe2/fb/dper/layer_models/tests/split_1/sparse_nn_test_4-3.7#binary,link-tree/caffe2/caffe2/fb/dper/layer_models/tests/split_1/sparse_nn_test_4.py", line 903, in test_cat_cat_embedding_arch_with_transfter_on_last
>     self.assertAlmostEqual(results["model_ne"], 0.1744275, delta=1e-5)
> AssertionError: 0.17446803612154085 != 0.1744275 within 1e-05 delta (4.053612154084063e-05 difference)

> ======================================================================
> FAIL: test_beta_loss (caffe2.caffe2.fb.dper.layer_models.tests.split_1.sparse_nn_test_4.SparseNNTest4)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/data/users/jongsoo/fbsource/fbcode/buck-out/dev/gen/caffe2/caffe2/fb/dper/layer_models/tests/split_1/sparse_nn_test_4-3.7#binary,link-tree/caffe2/caffe2/fb/dper/layer_models/tests/split_1/sparse_nn_test_4.py", line 681, in test_beta_loss
>     results["model_avg_lifetime_metric"], -1.958006, delta=1e-5
> AssertionError: -1.9943738633543253 != -1.958006 within 1e-05 delta (0.0363678633543254 difference)
```

Differential Revision: D18834383

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 0ce6a13d:


* **1/1** broken upstream at merge base a54dc87e (see [grid view](https://dr.pytorch.org/master-timeline.html?count=1&sha1=a54dc87e8ebf7634d3fb2f32dd32b73c1a4d095f&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
  <details>
  <summary>
  You may want to rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch origin viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch origin viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  
* **0/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=0ce6a13dd161ab8457b21f18012c770ce2320a09).

### :construction: 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably caused by upstream breakages:
* [pytorch_linux_xenial_py3_clang5_asan_test](https://dr.pytorch.org/build-details.html?build_id=78652388)



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

jspark1105 committed:
use avx2 and fma in adagrad (#229)

Summary:
Pull Request resolved: https://github.com/pytorch/FBGEMM/pull/229

Pull Request resolved: https://github.com/pytorch/pytorch/pull/30829

Using FMA should be faster and provide (slight) accuracy benefits.

Test Plan:
buck test caffe2/caffe2/fb/dper/layer_models/tests/split_1:sparse_nn_test_4

Without increasing the tolerance, we get the following errors and this may tell us that these tests are either numerically too sensitive or the tolerance range is too small. For test_beta_loss, I guess using fma actually improved the accuracy.

```
> ======================================================================
> FAIL: test_cat_cat_embedding_arch_with_transfter_on_last (caffe2.caffe2.fb.dper.layer_models.tests.split_1.sparse_nn_test_4.SparseNNTest4)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/data/users/jongsoo/fbsource/fbcode/buck-out/dev/gen/caffe2/caffe2/fb/dper/layer_models/tests/split_1/sparse_nn_test_4-3.7#binary,link-tree/caffe2/caffe2/fb/dper/layer_models/tests/split_1/sparse_nn_test_4.py", line 903, in test_cat_cat_embedding_arch_with_transfter_on_last
>     self.assertAlmostEqual(results["model_ne"], 0.1744275, delta=1e-5)
> AssertionError: 0.17446803612154085 != 0.1744275 within 1e-05 delta (4.053612154084063e-05 difference)

> ======================================================================
> FAIL: test_beta_loss (caffe2.caffe2.fb.dper.layer_models.tests.split_1.sparse_nn_test_4.SparseNNTest4)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/data/users/jongsoo/fbsource/fbcode/buck-out/dev/gen/caffe2/caffe2/fb/dper/layer_models/tests/split_1/sparse_nn_test_4-3.7#binary,link-tree/caffe2/caffe2/fb/dper/layer_models/tests/split_1/sparse_nn_test_4.py", line 681, in test_beta_loss
>     results["model_avg_lifetime_metric"], -1.958006, delta=1e-5
> AssertionError: -1.9943738633543253 != -1.958006 within 1e-05 delta (0.0363678633543254 difference)
```

Differential Revision: D18834383

fbshipit-source-id: bdf18732289ec37456ea46cc90381b7f0906f580
